django/db/models/deletion.py

```diff
@@ -247,7 +281,31 @@ def collect(self, objs, source=None, nullable=False, collect_related=True,
                 if hasattr(field, 'bulk_related_objects'):
                     # It's something like generic foreign key.
                     sub_objs = field.bulk_related_objects(new_objs, self.using)
-                    self.collect(sub_objs, source=model, nullable=True)
+                    self.collect(sub_objs, source=model, nullable=True, fail_on_restricted=False)
+
+        if fail_on_restricted:
+            # During collection we may have found objects via RESTRICT
+            # relationships (restricted objects). To be able to delete them, we
+            # need to also have found them through a different CASCADE
+            # relationship. Fail here if any of them wasn't.
```

Comment:
Please don't use `we` in comments per [Python style](https://docs.djangoproject.com/en/2.2/internals/contributing/writing-code/coding-style/#python-style), maybe:
```
# Raise an error if collected restricted objects (RESTRICT) aren't
# candidates for deletion, collected in the result of CASCADE.
```